### ISSUE TITLE
Rename the default file name generated when posting to /upload

### ISSUE THREAD

**Issue by YouhannaSabry** [strapi/strapi#1805]
**Informations**
- **Node.js v9.5.0**:
- **npm 6.1.0**:
- **Strapi 3.0.0-alpha.13.1**:
- **Database MongoDB**:
- **Ubuntu 18.04 Bionic Beaver**:

**What is the current behavior?**
Posting to /upload generates a name for the file and saves it to public/uploads .. I need to know how to customize or choose a specific name for the file being uplaoded.

**Comment by MartinMuzatko** [comment]
Would be nice to have this configurable

**Comment by dkarski** [comment]
Yes, would be nice becaouse now it isn't possible to change name until upload files

**Comment by lauriejim** [comment]
It will come with this feature https://portal.productboard.com/strapi/1-public-roadmap/c/28-media-library-with-edit-crop-resize-capabilities

**Comment by pljspahn** [comment]
I'm having a hard time understanding why anyone would want all their image filenames to just be random hash strings. The only benefit I see is that you avoid accidentally overwriting an existing file when uploading a new one with the same name (which might even be a drawback depending on the user - ie I _want_ that file to be overwritten so that my upload folder doesn't get polluted with duplicate files with different names).

I guess there must be something obvious I'm missing. 

It seems the only solution would to be build a different application with it's own endpoints that simply handles images, or to start mucking around in the strapi upload plugin to have it do what I want. Neither of those options please me.

Are there any solutions to this that don't involve building a separate application or modifying the plugin code manually?

**Comment by derrickmehaffy** [comment]
@pljspahn it's exactly as you said. If one person uploads "myblogheader01.png" then someone else uploads a different file with the same name what happens?

Either the first one is overwritten or the 2nd fails to upload. This is a common method in just about any upload provider.

**Comment by pljspahn** [comment]
It's pretty common to append a suffix when names are the same. Original file is myfile.jpg and the second gets appended to myfile02.jpg or something to avoid collisions. It's a very easy problem to avoid. Or put them in their own subfolder of uploads.

**Comment by derrickmehaffy** [comment]
@pljspahn you are more than welcome to use extensions to modify the code to do this. But as it stands I do not believe there is any plans to change this logic with Strapi proper.

**Comment by robots4life** [comment]
@derrickmehaffy Could you kindly help me getting uploaded images saved with the different breakpoints to have the following naming pattern please? This is Strapi 3.2.4. :rocket: 

```javascript
`Filename_Breakpoint_RandomString.ext`
```

I don't mind the random string and think it is good to have this. However it would be great to continue to have the filename at the beginning of the created files for each breakpoint.

Going over this https://github.com/sarpisik/strapi-customize-image-formats/tree/upload/extensions/upload/services from this recent tut https://sarpisik.com/blog/how-to-generate-different-image-formats-with-strapi-plugin-upload that speaks about how to modify things.

With this I am able to change the data that is stored in the Upload API however the file naming keeps having the breakpoint at the beginning of the file.

Do I need to modify something [from here](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/services/image-manipulation.js) to have different filenames when stored with breakpoints or is this done with code [from here](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/services/Upload.js)?

In particular I am having trouble following @lauriejim  's [suggestion](https://github.com/strapi/strapi/issues/1805#issuecomment-416167263) because the code updated since then.
> When you upload a file using `/upload` route, it call [this action](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/controllers/Upload.js#L12).
> Then at [this line](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/controllers/Upload.js#L35) strapi format files and define the `hash` that will became the file name (code [here](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/services/Upload.js#L46))
> 
> So you can customize the returned `hash` and it will change the name of your file.

Could you perhaps kindly point me at the right spots to look at in the code that are responsible for creating the final filename for the files with the chosen breakpoints? :thinking: 

All I am after is something like this or at least knowing a bit better where to look for the naming pattern, thank you. :+1: 
```javascript
`${filename}_${breakpoint}_${randomSuffix()}.${ext}`
```

**Comment by robots4life** [comment]
@derrickmehaffy I managed to [get the filename pattern](https://github.com/sarpisik/strapi-customize-image-formats/issues/1#issuecomment-715632392) I am after.

Am I right understanding that if the strapi-upload-plugin code changes in the future that these edits made will possibly have to adhere to those changes?

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy I managed to [get the filename pattern](https://github.com/sarpisik/strapi-customize-image-formats/issues/1#issuecomment-715632392) I am after.
> 
> Am I right understanding that if the strapi-upload-plugin code changes in the future that these edits made will possibly have to adhere to those changes?
> 
> edit
> From https://strapi.io/documentation/v3.x/concepts/customization.html#plugin-extensions
> So yeah, if I change the default behavior I need to check if that works on every update.
> ![Screenshot_20201024_141229](https://user-images.githubusercontent.com/34193243/97081480-fcaeec80-1602-11eb-80e1-fc336d698dee.png)

Yes that is correct, it's the downside to the extensions system, even minor versions could be breaking changes to you and you will have to keep your code in sync with ours.
